src/Neo/SmartContract/Native/NeoToken.cs

```diff
@@ -327,13 +327,35 @@ public BigInteger UnclaimedGas(DataCache snapshot, UInt160 account, uint end)
             return CalculateBonus(snapshot, state, end);
         }
 
+        [ContractMethod(Hardfork.HF_Echidna, RequiredCallFlags = CallFlags.States | CallFlags.AllowNotify)]
+        private async ContractTask OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash)
+                throw new InvalidOperationException("only GAS is accepted");
+
+            if ((long)amount != GetRegisterPrice(engine.SnapshotCache))
+                throw new ArgumentException("incorrect GAS amount for registration");
+
+            var pubkey = ECPoint.DecodePoint(data.GetSpan(), ECCurve.Secp256r1);
+
+            if (!RegisterInternal(engine, pubkey))
+                throw new InvalidOperationException("failed to register candidate");
+
+            await GAS.Burn(engine, Hash, amount);
+        }
+
         [ContractMethod(RequiredCallFlags = CallFlags.States)]
         private bool RegisterCandidate(ApplicationEngine engine, ECPoint pubkey)
         {
-            if (!engine.CheckWitnessInternal(Contract.CreateSignatureRedeemScript(pubkey).ToScriptHash()))
-                return false;
             // In the unit of datoshi, 1 datoshi = 1e-8 GAS
             engine.AddFee(GetRegisterPrice(engine.SnapshotCache));
```

Comment:
Sorry for nitpicking, but I've just noticed this issue: from the code point there's almost no difference, you've just moved the common code to a separate method. But execution result technically may be different (comparing the old and the new behaviour of `registerCandidate`): consider transaction that calls `registerCandidate` *and* has insufficient system fee *and* is not witnessed by the candidate:
* Old implementation will return `false` after witness check, and execution state will be HALT.
* New implementation will panic on `engine.AddFee`, and execution state will be FAULT.

So technically, it's a state diff. Practically, there likely no such transaction in the network (but we may always submit one!). Anyway, I'd suggest to get the old code back, to avoid possible state diff.

Reply:
Really nice catch. FAULT can be leveraged on its own, but with 1000 GAS paid you can even have different HALTs (the result of `System.Runtime.GasLeft` can tell the difference for the rest of the malicious script). This can be fixed by duplicating witness check in the `RegisterCandidate` (in a hardfork-dependent way), but I won't update this PR now since there are other ideas on how to solve this (which were not even discussed anywhere). So if that's not the "right" approach, I'll close the PR anyway.